bytecomp/matching.ml

```diff
@@ -2874,54 +3206,67 @@ let is_mutable p = find_in_pat have_mutable_field p
    2. And there are  guards or lazy patterns.
 *)
 
-let check_partial is_mutable is_lazy pat_act_list = function
+let check_partial is_mutable is_lazy pat_or_pats_act_list = function
   | Partial -> Partial
   | Total ->
       if
-        pat_act_list = [] ||  (* allow empty case list *)
+        pat_or_pats_act_list = [] ||  (* allow empty case list *)
         List.exists
-          (fun (pats, lam) ->
-            is_mutable pats && (is_guarded lam || is_lazy pats))
-          pat_act_list
+          (fun (pat_or_pats, lam, _loc) ->
+            is_mutable pat_or_pats && (is_guarded lam || is_lazy pat_or_pats))
+          pat_or_pats_act_list
       then Partial
       else Total
 
-let check_partial_list =
-  check_partial (List.exists is_mutable) (List.exists is_lazy)
-let check_partial = check_partial is_mutable is_lazy
+let check_partial_list (pats_act_list : pats_act_list) =
+  check_partial (List.exists is_mutable) (List.exists is_lazy) pats_act_list
+
+let check_partial (pat_act_list : pat_act_list) =
+  check_partial is_mutable is_lazy pat_act_list
 
 (* have toplevel handler when appropriate *)
 
-let start_ctx n = [{left=[] ; right = omegas n}]
+let start_ctx loc n = [{left=[] ; right = omegas loc n}]
 
-let check_total total lambda i handler_fun =
+let check_total loc total lambda i handler_fun =
   if jumps_is_empty total then
     lambda
   else begin
-    Lstaticcatch(lambda, (i,[]), handler_fun())
+    Lstaticcatch(lambda, (i,[]), handler_fun(), loc)
   end
 
-let compile_matching repr handler_fun arg pat_act_list partial =
+let pats_act_list_of_pat_act_list (pat_act_list : pat_act_list) =
```

Comment:
What exactly is the complaint here?

Reply:
Well the function name `pats_act_list_of_pat_act_list` is not very pleasant to read on the caller side.

Maybe we could find better formulations for your types `pats_act_list` and `pat_act_list`.

In general I'm not convinced by `foo_list`, why not use the more flexible `foo list` instead? So we could find a name for "a single pattern and the action to take" and "many parallel patterns and the action to take". Suggestions:
- `simple_clause` versus `multi_clause` ("clause" means "something with the shape `foo -> bar`)
- `pattern_action` versus `row_action`
- `pattern_clause` versus `row_clause`
- `clause1` versus `clause` (urgh)

Note that for the function itself you could just write `row_clause_of_pattern_clause` and then do `List.map` on it at the callsite.

Reply:
Well, it's `foo_list` because I think the whole data structure should have a single name; it's just I didn't choose a good one.  If you and/or @trefis agree on a name, I will gladly change it.

Reply:
(To clarify, I think the naming should reflect what the thing is, not what structure is being used to represent it.)

Reply:
Another thing that may work is `type 'a case = ('a * lambda * Location.t)` (why not `'a * lambda Location.loc`?), so `pattern case` and `pattern list case`, with `type 'a cases = 'a case list`.

Reply:
> Well the function name `pats_act_list_of_pat_act_list` is not very pleasant to read on the caller side.

Or at the definition point.

I like the `'a case` proposition.
In the end, I'd go with the following definitions:
```ocaml
type row = pattern list
type 'a case = 'a * lambda * Location.t

(* ... *)

let mk_row_case (pat, act, loc) : row case =  ([pat], act, loc)
```
And then just `List.map mk_row_case` (I'm not a fan of that name, but it's less painful than the previous one) on the few places that need it.

Reply:
I've dealt with all of this.